### PR TITLE
Sequence to sequence translation

### PR DESCRIPTION
**Pull request by peastman**
There's still a lot to do on this, both in terms of features and documentation, but I'd like to start getting feedback on it.

### CODE REVIEW

File: deepchem/models/tensorgraph/models/seqtoseq.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+          embedding_dimension, self.batch_size, in_layers=prev_layer)
+    output_layer = layers.Dense(
+        len(output_tokens), in_layers=prev_layer, activation_fn=tf.nn.softmax)
+    self.add_output(output_layer)
+    prob = layers.ReduceSum(
+        layers.Multiply([output_layer, self._labels]), axis=2)
+    log_prob = layers.Log(
+        layers.Add([prob, layers.Constant(np.finfo(np.float32).eps)]))
+    objective = layers.ReduceMean(layers.ReduceSum(log_prob, axis=1))
+    loss = layers.Multiply([objective, layers.Constant(-1)])
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/seqtoseq.py]: Should we swap this math over to the new operator syntax?

File: deepchem/models/tensorgraph/models/seqtoseq.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    self.set_loss(loss)
+    self.output = output_layer
+
+  def fit_sequences(self,
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/seqtoseq.py]: We might not want a new name here; we could just reuse `fit_generator` since the only meaningful inputs for `Seq2Seq` is sequences.

File: deepchem/models/tensorgraph/models/seqtoseq.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/tensorgraph/models/seqtoseq.py]: Already done.  That PR hadn't been merged yet when I wrote this.

File: deepchem/models/tensorgraph/models/seqtoseq.py (hunk 2 above)
Comment (peastman) [reply deepchem/models/tensorgraph/models/seqtoseq.py]: I'd like to keep open the option of calling `fit_generator()`.  Note that although the two methods both taking generators, they're very different.  This one takes a generator that produces pairs of sequences.  The other takes a generator that produces feed dicts.

File: deepchem/models/tensorgraph/models/seqtoseq.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        if token == SeqToSeq.sequence_end:
+          break
+        s.append(token)
+      return s
+
+    # Do a beam search with length normalization.
+
+    logprobs = np.log(probs)
+    # Represent each candidate as (normalized prob, raw prob, sequence)
+    candidates = [(0.0, 0.0, [])]
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/models/seqtoseq.py]: maybe move this into a different function?  Looking through the most likely beams and their normalized prob is super useful.

Also impressive implementation.

**Comment by peastman** [comment]
I've added support for using `SeqToSeq` as a variational autoencoder.  But I've been unable to get it to work properly in any of my tests, so I'm hoping someone with more experience in VAEs can maybe help.

When used as a VAE, it adds random noise to the embedding vector, and the loss is a sum of two terms: the standard reconstruction loss, and the KL divergence of the embedding distribution from a unit Gaussian.  That second term ends up completely dominating the optimization.  It quickly drives the second term to a tiny value, but never manages to optimize the first term at all.  That sounds to me like I just need to change their relative magnitudes.  And in fact, if I multiply the KL divergence by 0.1, then it is able to learn.  It learns more slowly than normally, and never manages to do as well, but it does get to the point where it can reproduce a significant fraction of input sequences.

But in every source I've looked at, no one ever reduces the scale of that term.  And in fact some of them imply that the scale is not arbitrary, and there are good reasons for it having exactly the magnitude it does.  So is there something else I'm doing wrong?

**Comment by rbharath** [comment]
@peastman Perhaps try taking a look at vae_loss in https://github.com/maxhodak/keras-molecules/blob/master/molecules/model.py and copying their scaling terms for the cross entropy and KL terms?

**Comment by peastman** [comment]
They don't apply any scaling.  Of course, that's a somewhat different model, since their encoder is a CNN instead of an RNN.

I was hoping someone had experience with VAEs and might understand what's going on?  It seems to me that the two terms are on fundamentally different scales, since the reconstruction loss is proportional to the sequence length, while the KL loss is proportional to the embedding vector length, and there's no fixed relationship between those.  But perhaps there's more to it that I don't understand.

**Comment by rbharath** [comment]
Unfortunately, haven't yet trained a VAE myself :). Glad to poke around with it and will ping this thread if I figure anything useful out

**Comment by rbharath** [comment]
Potentially relevant: https://github.com/cheng6076/Variational-LSTM-Autoencoder

**Comment by peastman** [comment]
I've looked at lots of VAE implementations, and blog posts, and papers...  Here's the one I copied the equations from: http://kvfrans.com/variational-autoencoders-explained/

**Comment by peastman** [comment]
Once the tests pass, I think this is ready to merge.  I found a solution to the variational autoencoder problem in https://arxiv.org/pdf/1511.06349.pdf.

**Comment by rbharath** [comment]
@peastman This looks awesome!

LGTM
